File: libraries/ecore/src/ecore/NotificationChain.cpp
// *****************************************************************************
//
// This file is part of a MASA library or program.
// Refer to the included end-user license agreement for restrictions.
//
// Copyright (c) 2018 MASA Group
//
// *****************************************************************************

#include "ecore/NotificationChain.hpp"
#include "ecore/ENotification.hpp"
#include "ecore/ENotifier.hpp"

using namespace ecore;

bool NotificationChain::add( const std::shared_ptr<ENotification>& notification )
{
    if (!notification )
        return false;
    for( auto n : notifications_ )
    {
        if( n && n->merge( notification ) )
            return false;
    }
    notifications_.push_back( notification );
    return true;
}

void NotificationChain::dispatch()
{
    for( auto notification : notifications_ )
    {
        auto notifier = notification->getNotifier();
        if( notifier &&  notification->getEventType() != -1 )
            notifier->eNotify( notification );
    }

}

File: ecore/go/ecore/notificationchain.go
package ecore

// notificationChain is an implementation of ENotificationChain interface
type notificationChain struct {
	notifications *arrayEList
}

// NewNotificationChain ...
func NewNotificationChain() *notificationChain {
	return &notificationChain{notifications: NewEmptyArrayEList()}
}

// Add Adds a notification to the chain.
func (chain *notificationChain) Add(newNotif ENotification) bool {
	if newNotif == nil {
		return false
	}
	for it := chain.notifications.Iterate(); it.Next(); {
		if it.Value().(ENotification).Merge(newNotif) {
			return false
		}
	}
	chain.notifications.Add(newNotif)
	return true
}

// Dispatch Dispatches each notification to the appropriate notifier via notifier.ENotify method
func (chain *notificationChain) Dispatch() {
	for it := chain.notifications.Iterate(); it.Next(); {
		value := it.Value().(ENotification)
		notifier := value.GetNotifier()
		if notifier != nil && value.GetEventType() != -1 {
			notifier.ENotify(value)
		}
	}
}